src/solver/opamCudf.ml

```diff
@@ -1037,35 +1028,129 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
                   Printf.sprintf "%s: no longer available"
                     (OpamPackage.to_string (cudf2opam p))
                 in
-                `Missing (Some csp, msg, OpamFormula.Empty)
+                let csp = cst_to_string csp in
+                `NoLongerAvailable (csp, msg, OpamFormula.Empty)
               else
               let fdeps = formula_of_vpkgl cudfnv2opam packages deps in
-              let sdeps = OpamFormula.to_string fdeps in
-              `Missing (Some csp, sdeps, fdeps)
+              `Missing (csp, fdeps)
             in
-            if List.mem msg explanations then raise Not_found else
-              msg :: explanations, ct_chains
+            msg :: explanations
           | Dependency _ ->
-            explanations, ct_chains
+            explanations
         with Not_found ->
-          explanations, ct_chains)
-      ([], ct_chains) reasons
+          explanations)
+      [] reasons
+  in
+
+  let rec simplify_formula formula1 formula2 =
+    let open OpamStd.Option.Op in
+    match formula1, formula2 with
+    | Empty, Empty -> Some Empty
+    | Atom (name1, vformula1), Atom (name2, vformula2) ->
+      if OpamPackage.Name.equal name1 name2 then
+        OpamFormula.simplify_version_formula (Or (vformula1, vformula2))
+        >>= fun vformula -> Some (Atom (name1, vformula))
+      else
+        None
+    | Block x, y | x, Block y -> simplify_formula x y
+    | And (x1, x2), And (y1, y2) ->
+        simplify_formula x1 y1 >>= fun z1 ->
+        simplify_formula x2 y2 >>= fun z2 ->
+        Some (And (z1, z2))
+    | Or (x1, x2), Or (y1, y2) ->
+        simplify_formula x1 y1 >>= fun z1 ->
+        simplify_formula x2 y2 >>= fun z2 ->
+        Some (Or (z1, z2))
+    | Empty, _
+    | Atom _, _
+    | And _, _
+    | Or _, _ -> None
+  in
+
+  let explanations =
```

Comment:
It may worth adding some comment of the explanations constructions, as there is several steps to have the final explanation list?

Reply:
added in https://github.com/ocaml/opam/pull/6106/commits/b591657225fe9c5b5904788e3c107b567470c89d. What do you think?